### ISSUE TITLE
Distinguishing MatchParen from Cursor

### ISSUE THREAD

**Issue by schreifels** [nanotech/jellybeans.vim#71]
Thanks for this fantastic color scheme! I've been using it for quite some time, and there's one thing that has bugged me for a while. Consider this file:

<img width="374" alt="normal" src="https://user-images.githubusercontent.com/1530012/31646400-916cdaac-b2b6-11e7-89ec-ffa8a8b85f86.png">

If I put the cursor over the opening `{`, it looks like this:

<img width="376" alt="before1" src="https://user-images.githubusercontent.com/1530012/31646398-913a5aaa-b2b6-11e7-8d1d-6f91f134a80a.png">

And over the closing `}`:

<img width="373" alt="before2" src="https://user-images.githubusercontent.com/1530012/31646399-91524048-b2b6-11e7-949c-bdae82401e9e.png">

If I look carefully, I can see the color difference, but I find it very difficult to distinguish between them when working with less contrived code in my day-to-day. Using the same color as Function (it's intended to take the fg color of `{` and make it the bg color, but I'm not sure that I got the correct group), I applied this change to my local config:

```vim
let g:jellybeans_overrides = { 'MatchParen': { 'guifg': '000000', 'guibg': 'fad07a' } }
```

which looks like this when the cursor is on `{`:

<img width="377" alt="after1" src="https://user-images.githubusercontent.com/1530012/31646396-910d12ca-b2b6-11e7-9cbb-c808a5c38f42.png">

and this on `}`:

<img width="376" alt="after2" src="https://user-images.githubusercontent.com/1530012/31646397-9123e0cc-b2b6-11e7-8688-996cd0fcc6dc.png">

I was wondering if this (or some other change to the same effect) might be appropriate to make in the core color scheme. Thanks!

**Comment by nanotech** [comment]
Thanks for the report. It looks like I've been using a non-default cursor color in my terminal for so long that I've never noticed this.

Here's how I intend the cursor to appear, as it does in graphical Vim (MacVim):

![macvim-blink](https://user-images.githubusercontent.com/34699/31806868-1175df64-b528-11e7-9bc3-1daecbf6b184.gif)

and here's how it appears for me in Terminal.app with a custom cursor color (iTerm looks basically identical with the same cursor color set):

![terminal-blink](https://user-images.githubusercontent.com/34699/31806871-1411c454-b528-11e7-9b21-5e4fa54f5580.gif)

Vim isn't able to change the terminal cursor color as far as I can tell, so I agree that MatchParen should have a default that's distinguishable from common default cursor colors. Of course, various terminals have all chosen different colors. A quick survey:

- Terminal.app: `#5f5f5f`
- iTerm: Smart Cursor Color (whiteish when on a dark background)
- gnome-terminal on Ubuntu: Reverse fg/bg (white when on MatchParen)
- xterm on Ubuntu: `#e5e5e5`

Some possible solutions:

- Detect Terminal.app, assume the Pro theme is being used, and change MatchParen only there. (Will be wrong if the user is not using Pro or changes their cursor color.)
- Add a note to the readme asking Terminal.app users to change their cursor color. (Will be missed by most existing and new users.)
- Change MatchParen to some color that is distinguishable from both gray and white, as you suggested. (Will either differ from the graphical version or will change both.)

**Comment by schreifels** [comment]
Thanks for looking into this, @nanotech ! Given how long this has gone unreported, I suspect either not many people are using my setup or it just doesn't bother them. Doing something specific for Terminal.app seems suboptimal for the reasons you mentioned.

Personally, I would advocate for the 3rd solution since this would resolve the issue across the board without any special case logic, but I think the 2nd solution also makes sense if you are attached to how MatchParen is colored in terminals other than Terminal.app.

**Comment by nanotech** [comment]
I chose: all of the above!

Trying different colors, I found using any `MatchParen` background color brighter than the cursor would get me confused about which side was the cursor.  With the lower cursor brightness range being so low, that leaves a bright foreground on a dark background. Magenta on black seemed acceptably visible.

![terminal-blink2](https://user-images.githubusercontent.com/34699/32687354-cc1db748-c67f-11e7-873d-75e02d07b558.gif)

Bright on dark also works better for 16-color terminals that don't have a "grey as ANSI black" palette configured.

I do prefer the existing colors though, and don't want to unnecessarily change the colorscheme for existing users, so I've special cased the new alternate colors to only appear by default for Terminal.app and 16-color terminals. A bright on dark `MatchParen` will be distinguishable from any reasonable cursor color on a dark background even if the user changes it.

For those using Terminal.app with bright cursors or 16-color terminals with custom palettes, I've added override examples to the readme.
